File: Src/ButterflyNet.Parser/Strategies/EmphasisStrategy.cs
using System;
using ButterflyNet.Parser.Satisfiers;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {

	[TokenTransformer("''")]
	public abstract class EmphasisStrategy : InlineStrategy {
		protected Type Type { get { return ScopeTypeCache.Emphasis; } }
	}

	public class OpenEmphasisStrategy : EmphasisStrategy {
		public OpenEmphasisStrategy() {
			AddSatisfier(new OpenNonNestableInlineScopeSatisfier(Type));
		}

		protected override void DoExecute(ParseContext context) {
			OpenScope(new EmphasisScope(), context);
		}
	}

	public class CloseEmphasisStrategy : EmphasisStrategy {
		public CloseEmphasisStrategy() {
			AddSatisfier(new CurrentScopeMustMatchSatisfier(Type));
		}

		protected override void DoExecute(ParseContext context) {
			CloseCurrentScope(context);
		}
	}
}

File: src/js/src/strategies/EmphasisStrategy.js
function EmphasisStrategy() {
	EmphasisStrategy.$parent.call(this);
	this.type = ScopeTypeCache.emphasis;
	this.setAsTokenTransformer("''");
}

extend(InlineStrategy, EmphasisStrategy);

function OpenEmphasisStrategy() {
	OpenEmphasisStrategy.$parent.call(this);
	
	this.addSatisfier(new OpenNonNestableInlineScopeSatisfier(this.type));
	
	this.doExecute = function(context) {
		this.openScope(new EmphasisScope(), context);
	};
}

extend(EmphasisStrategy, OpenEmphasisStrategy);

function CloseEmphasisStrategy() {
	CloseEmphasisStrategy.$parent.call(this);
	
	this.addSatisfier(new CurrentScopeMustMatchSatisfier(this.type));
	
	this.doExecute = function(context) {
		this.closeCurrentScope(context);
	};
}

extend(EmphasisStrategy, CloseEmphasisStrategy);